### PR TITLE
Move the psid-map to softwire in snabb-softwire-v2

### PR DESCRIPTION
**Pull request by tsyesika**
This is the implementation #792.

This PR moves the `psid-map` list in the yang schema over to the `softwire` list. There were many options or routes to implementation however I tried to pick the ones which I thought would keep the code the simplest, fastest with the smallest size footprint with regard to the binding table.

The new softwire schema looks like this:
```yang
list softwire {
   key "ipv4 psid padding";

   leaf ipv4 {
      type inet:ipv4-address;
      mandatory true;
   }

   leaf padding {
      type uint16 { range 0..0; }
      default 0;
   }

   leaf br-address {
      type inet:ipv6-address;
      mandatory true;
   }

   leaf b4-ipv6 {
      type inet:ipv6-address;
      mandatory true;
   }

   leaf psid {
      type uint16;
      mandatory true;
   }

   container port-set {
      leaf psid-length {
        type uint8 { range 0..16; }
        mandatory true;
      }

      leaf shift {
        type uint8 { range 0..16; }
      }

      leaf reserved-ports-bit-count {
      type uint8 { range 0..16; }
      default 0;
      }
   }
}
```
_NB: I've removed the descriptions above but not in the schema itself to make it more readable._

I did want to put the `psid` field in the `port-set` container as I think it makes the most sense however, YANG requires all keys be children of the `list` (section: [7.8.2](https://tools.ietf.org/html/rfc6020#section-7.8.2)).

I decided I'd keep `binding_table.lua` using the separated `psid-map` and `softwire` lists in order to save space and retain the ability to lookup with a ctable via the streaming interface. Changing the schema to the above means: the `softwire` list changes from a ctable to a cltable with the value being a simple lua table. Maybe shift should be `default 0;` and I believe that should prevent that however it wasn't in the original schema (v1). The ctable is desired and I also **don't** want the psid information on the softwire as it's already going to be in the `psid-map`.

The change also means that the `end-addr` option goes away! This also brings up the option of switching away from `rangemap.lua` to a plain-old ctable. I decided to take this route in order to allow me to add a new `psid-map` entry without restarting. That however changes some of the interfaces for example the `lookup` interface.

So what actually happens with these changes?

### (Re-)Starting the lwaftr

In [load](src/apps/lwaftr/binding_table.lua) I iterate over the softwires to build up the `psid-map` ctable and put the softwires in a ctable without the now superfluous information.

### Adding a softwire via the reconfigurable interface

1. We see it's a softwire and hand off to [add_softwire_entry_actions](src/apps/config/support/snabb-softwire-v2.lua) as per usual
2. We create a blob for the ctable
3. We check if there is a psid-map entry for that ctable
4. If not we also make an rpc to our new shiny `add_psid_map_entry` function.
5. We commit those changes as per usual and the binding table adds it to its ctable as per usual too (adding `psid-map` entries is done in a similar fashion to adding softwires).

**Question:** What do we do if they add a softwire with an existing `psid-map` entry however the `psid-map` entry is different from others with the same IP? Do we update the `psid-map` entry changing it for the other softwires or error?

What's left to do is:
- [x] As the `load` function has changed since I wrote and tested the migration, I need to fix that.
- [x] Update the translation layer
- [x] Update the documentation.

### CODE REVIEW

File: src/program/lwaftr/tests/data/tunnel_icmp.conf [hunk 1]
```diff
@@ -1,101 +1,124 @@
...
-      psid-length 6;
-      shift 10;
+    softwire {
+      ipv4 178.79.150.3;
+      psid 4;
+      b4-ipv6 127:14:25:36:47:58:69:128;
+      br-address 1e:2:2:2:2:2:2:af;
+      port-set {
+        psid-length 0;
+        reserved-ports-bit-count 16;
```
Comment (wingo) [inline src/program/lwaftr/tests/data/tunnel_icmp.conf]: AFAIU this initial config always has psid-length+ shift = 16 so reserved-ports-bit-count should be 0; no?  This comment applies here and elsewhere.

**Review by wingo — COMMENTED** [review]
On the low level: excellent, excellent work.  Nice stuff.  Just a nit regarding what is probably an error in the migrate-configuration thing.

On the high level, have you done any performance testing yet?  I am a little skeptical that there will be an advantage to using a ctable internally for the psid map.  To me it seems like a range map is the right thing; dunno.  Wdyt?  Is there a reason to switch?

**Comment by tsyesika** [comment]
I have not yet done performance testing, it's my next job, hopefully shouldn't take long.

The reason for switching psidmap to a ctable over a rangemap is simply so we can add new psid entries without having to rebuild it. It becomes quite difficult to add them otherwise. It seems the rangemap was mainly useful because it allowed for range lookup was it not? We can switch back but I personally think being able to add or remove a psid-map entry is quite valuable however maybe it's not worth it.

**Comment by wingo** [comment]
right, you do have to rebuild the table when you add new ips to it.  that happens infrequently though.  relatedly it's hard to know when to rebuild:
 - when you remove a softwire, do you leave its entry in the psid map?
   - if you remove the psid map entry, you could corrupt other softwires on that ip
   - if you keep it you could mistakenly identify that ip as yours for the purpose of hairpinning
   - when you add a softwire and the psid map has a conflicting entry, do you know for sure that that entry is used?

We could add refcounts or something in the psid map but for now I think it's OK to assume that modifying the PSID map is a bit of a PITA and so the data structure change is the least of our worries.

On the perf side I am wary of introducing the ctable overhead here.  Pretty sure the binary search in the rangemap is a win.  I guess you can go ahead with perf testing of this branch as-is and then we can see what the real-world win (if any) of the rangemap is -- but that would definitely be my first suggestion as a perf improvement.

**Comment by tsyesika** [comment]
@wingo PTAL.

I've made some changes as you suggested, switching back to the rangemap helped a lot with performance issues I discovered last week. It does mean we restart if you do happen to add a softwire which defines a new port-set however hopefully that won't be too often. The `padding` parameter still exists too, this is largely related to https://github.com/Igalia/snabb/blob/lwaftr/src/lib/multi_copy.dasl#L57.

I have performance tested this and it can reach 2 MPPS with only 0.000050% packet loss, the interesting parts of the test are as follows:
```
Applying 9.000000 Gbps of load.
  NIC 0:
    TX 1959917 packets (1.959917 MPPS), 1077954350 bytes (8.999939 Gbps)
    RX 1959917 packets (1.959917 MPPS), 1156351030 bytes (9.627112 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  NIC 1:
    TX 1959917 packets (1.959917 MPPS), 1077954350 bytes (8.999939 Gbps)
    RX 1959917 packets (1.959917 MPPS), 999557670 bytes (8.372765 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.200000 Gbps of load.
  NIC 0:
    TX 2003477 packets (2.003477 MPPS), 1101912350 bytes (9.199966 Gbps)
    RX 2003476 packets (2.003476 MPPS), 1182050840 bytes (9.841074 Gbps)
    Loss: 0 ingress drop + 1 packets lost (0.000050%)
  NIC 1:
    TX 2003477 packets (2.003477 MPPS), 1101912350 bytes (9.199966 Gbps)
    RX 2003477 packets (2.003477 MPPS), 1021773270 bytes (8.558854 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.400000 Gbps of load.
  NIC 0:
    TX 2047038 packets (2.047038 MPPS), 1125870900 bytes (9.399998 Gbps)
    RX 2014834 packets (2.014834 MPPS), 1188752060 bytes (9.896865 Gbps)
    Loss: 0 ingress drop + 32204 packets lost (1.573200%)
  NIC 1:
    TX 2047038 packets (2.047038 MPPS), 1125870900 bytes (9.399998 Gbps)
    RX 2024271 packets (2.024271 MPPS), 1032378210 bytes (8.647686 Gbps)
    Loss: 0 ingress drop + 22767 packets lost (1.112192%)
Applying 9.600000 Gbps of load.
  NIC 0:
    TX 2090550 packets (2.090550 MPPS), 1149802500 bytes (9.599806 Gbps)
    RX 2037774 packets (2.037774 MPPS), 1202286660 bytes (10.009546 Gbps)
    Loss: 0 ingress drop + 52776 packets lost (2.524503%)
  NIC 1:
    TX 2090550 packets (2.090550 MPPS), 1149802500 bytes (9.599806 Gbps)
    RX 2090550 packets (2.090550 MPPS), 1066180500 bytes (8.930830 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.800000 Gbps of load.
  NIC 0:
    TX 2134151 packets (2.134151 MPPS), 1173783050 bytes (9.800021 Gbps)
    RX 2037836 packets (2.037836 MPPS), 1202323240 bytes (10.009850 Gbps)
    Loss: 0 ingress drop + 96315 packets lost (4.513036%)
  NIC 1:
    TX 2134151 packets (2.134151 MPPS), 1173783050 bytes (9.800021 Gbps)
    RX 2134151 packets (2.134151 MPPS), 1088417010 bytes (9.117093 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 10.000000 Gbps of load.
  NIC 0:
    TX 2177656 packets (2.177656 MPPS), 1197710800 bytes (9.999796 Gbps)
    RX 2038025 packets (2.038025 MPPS), 1202434750 bytes (10.010779 Gbps)
    Loss: 0 ingress drop + 139631 packets lost (6.411986%)
  NIC 1:
    TX 2177656 packets (2.177656 MPPS), 1197710800 bytes (9.999796 Gbps)
    RX 2177656 packets (2.177656 MPPS), 1110604560 bytes (9.302946 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 10.000000 Gbps of load.
  NIC 0:
    TX 2177658 packets (2.177658 MPPS), 1197711900 bytes (9.999806 Gbps)
    RX 2037978 packets (2.037978 MPPS), 1202407020 bytes (10.010548 Gbps)
    Loss: 0 ingress drop + 139680 packets lost (6.414230%)
  NIC 1:
    TX 2177658 packets (2.177658 MPPS), 1197711900 bytes (9.999806 Gbps)
    RX 2177658 packets (2.177658 MPPS), 1110605580 bytes (9.302955 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.800000 Gbps of load.
  NIC 0:
    TX 2134114 packets (2.134114 MPPS), 1173762700 bytes (9.799851 Gbps)
    RX 2037801 packets (2.037801 MPPS), 1202302590 bytes (10.009679 Gbps)
    Loss: 0 ingress drop + 96313 packets lost (4.513020%)
  NIC 1:
    TX 2134114 packets (2.134114 MPPS), 1173762700 bytes (9.799851 Gbps)
    RX 2134114 packets (2.134114 MPPS), 1088398140 bytes (9.116935 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.600000 Gbps of load.
  NIC 0:
    TX 2090593 packets (2.090593 MPPS), 1149826150 bytes (9.600003 Gbps)
    RX 2037801 packets (2.037801 MPPS), 1202302590 bytes (10.009679 Gbps)
    Loss: 0 ingress drop + 52792 packets lost (2.525217%)
  NIC 1:
    TX 2090593 packets (2.090593 MPPS), 1149826150 bytes (9.600003 Gbps)
    RX 2090593 packets (2.090593 MPPS), 1066202430 bytes (8.931013 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.400000 Gbps of load.
  NIC 0:
    TX 2047035 packets (2.047035 MPPS), 1125869250 bytes (9.399985 Gbps)
    RX 2037809 packets (2.037809 MPPS), 1202307310 bytes (10.009718 Gbps)
    Loss: 0 ingress drop + 9226 packets lost (0.450701%)
  NIC 1:
    TX 2047035 packets (2.047035 MPPS), 1125869250 bytes (9.399985 Gbps)
    RX 2047035 packets (2.047035 MPPS), 1043987850 bytes (8.744934 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.200000 Gbps of load.
  NIC 0:
    TX 2003480 packets (2.003480 MPPS), 1101914000 bytes (9.199980 Gbps)
    RX 2003480 packets (2.003480 MPPS), 1182053200 bytes (9.841094 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  NIC 1:
    TX 2003480 packets (2.003480 MPPS), 1101914000 bytes (9.199980 Gbps)
    RX 2003480 packets (2.003480 MPPS), 1021774800 bytes (8.558867 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
```

**Comment by wingo** [comment]
Perf numbers look good.  I have a few questions:
 1. What workload are you running -- 550-byte packets?
 2. How many softwires in your test binding table?  Probably we need to test with 2M.
 3. How many active softwires in the test workload?  20K is good enough for final perf numbers.
 4. What is the difference in the compiled configuration size before and after with a 2M-entry binding table?

It also looks to me like a few side utilities need to be updated, like the binding table generator. (`snabb lwaftr generate-binding-table`)

File: src/apps/lwaftr/binding_table.lua [hunk 2]
```diff
@@ -63,21 +63,18 @@ module(..., package.seeall)
 local bit = require('bit')
 local ffi = require("ffi")
 local rangemap = require("apps.lwaftr.rangemap")
-local cltable = require("lib.cltable")
+local ctable = require("lib.ctable")
+local ipv6 = require("lib.protocol.ipv6")
 
 local band, lshift, rshift = bit.band, bit.lshift, bit.rshift
 
 -- FIXME: Pull these types from the yang model, not out of thin air.
```
Comment (wingo) [inline src/apps/lwaftr/binding_table.lua]: We can remove this comment, given that the psid map is no longer a YANG thing.

File: src/apps/lwaftr/binding_table.lua [hunk 3]
```diff
@@ -195,11 +192,11 @@ end
...
 --
---   for ipv4_lo, ipv4_hi, psid_info in bt:iterate_psid_map() do ... end
+--   for entry in bt:iterate_psid_map() do ... end
 --
--- The IPv4 values are host-endianness uint32 values, and are an
--- inclusive range to which the psid_info applies.  The psid_info is a
--- psid_map_value_t pointer, which has psid_length and shift members.
+-- The IPv4 values are host-endianness uint32 values. The key is a
+-- softwire_key_t which has a "addr" and psid_info is a psid_map_value_t
+-- pointer, which has psid_length and shift members.
```
Comment (wingo) [inline src/apps/lwaftr/binding_table.lua]: I think the old comment still applies -- the psid map is a rangemap that will coalesce contiguous ranges of keys that have the same value.  However I think perhaps this function is no longer needed, as I think right now the only user of this function was to write out the psid-map yang list, and that is no more.  Consider removing the function and comment.

File: src/apps/lwaftr/lwaftr.lua [hunk 4]
```diff
@@ -276,6 +276,12 @@ end
 function LwAftr:remove_softwire_entry(entry_key_blob)
    self.binding_table:remove_softwire_entry(entry_key_blob)
 end
+function LwAftr:add_psid_map_entry(psid_blob)
```
Comment (wingo) [inline src/apps/lwaftr/lwaftr.lua]: What's this for?  Appears to not be called.

File: src/apps/lwaftr/rangemap.lua [hunk 5]
```diff
@@ -83,7 +83,7 @@ function RangeMapBuilder:add(key, value)
    self:add_range(key, key, value)
 end
 
-function RangeMapBuilder:build(default_value)
+function RangeMapBuilder:build(default_value, ignore_duplicates)
```
Comment (wingo) [inline src/apps/lwaftr/rangemap.lua]: I am pretty sure that we don't want to ignore duplicates -- rather we want to require that all entries with the same key have the same value.  Maybe that means the builder should use a ctable as its intermediate data store and verify that when a key is added but already present in the builder, that the value is the same.

File: src/lib/yang/data.lua [hunk 6]
```diff
@@ -349,7 +349,11 @@ local function table_parser(keyword, keys, values, string_key, key_ctype,
       end
       for k,_ in pairs(values) do
          local id = normalize_id(k)
-         value[id] = struct[id]
+         if type(struct[id]) == "cdata" and type(value[id]) == "cdata" then
```
Comment (wingo) [inline src/lib/yang/data.lua]: This seems like the wrong thing but I'm not sure.  Is this change needed?

File: src/program/lwaftr/doc/configuration.md [hunk 7]
```diff
@@ -134,47 +142,42 @@ encapsulated in IPv6 and sent to the AFTR; the AFTR does the reverse.
...
-    shift 10;
+  softwire {
+    ipv4 178.79.150.233;
+    psid 80;
+    b4-ipv6 127:2:3:4:5:6:7:128;
+    br-address 8:9:a:b:c:d:e:f
+    port-set {
+      psid-length 6;
+      reserved-ports-bit-count 0;
+      shift 10;
```
Comment (wingo) [inline src/program/lwaftr/doc/configuration.md]: I think this example is out of date; no "shift" parameter any more.

File: src/program/lwaftr/doc/configuration.md [hunk 8]
```diff
@@ -134,47 +142,42 @@ encapsulated in IPv6 and sent to the AFTR; the AFTR does the reverse.
...
+    }
   }
 ```
 
-`end-addr`, `reserved-ports-bit-count`, and `shift` are all optional.  
+`reserved-ports-bit-count` and `shift` in the `port-set` container are
+both optional.
 
 An entry's `psid-length`, `shift` and `reserved-ports-bit-count`
```
Comment (wingo) [inline src/program/lwaftr/doc/configuration.md]: Relatedly, this paragraph needs updating.

**Review by wingo — COMMENTED** [review]
Some nits.  Looking very good, thank you for your work here!

**Comment by tsyesika** [comment]
So, in answer to your questions:

> 1. What workload are you running -- 550-byte packets?

Yep, I am.

> 2. How many softwires in your test binding table? Probably we need to test with 2M.

I used https://people.igalia.com/dpino/lwaftr/lwaftr-migrated.conf which by my count has 1,000,062 softwires. Is there a configuration with 2M that I can use?

> 3. How many active softwires in the test workload? 20K is good enough for final perf numbers.

Unsure, Is there an easy way of getting the count?

> 4. What is the difference in the compiled configuration size before and after with a 2M-entry binding table?

This is using the above binding table I mentioned in your 2nd question of ~1M softwires.

Before: 93M
After: 166M

So it grew by 73M.

**Comment by wingo** [comment]
Thanks for the perf answers!  In order:

 1. Great.
 2. 1M is fine.  Just making sure it wasn't a test with just like 10 softwires.
 3. I don't know.  Open the pcap in wireshark and see how many packets are in it?  Or tcpdump?  I think these files are 20K active softwires though.
 4. Hmm.  I guess it's OK though given that our perf numbers are stable on this workload.

I think I have a fix for the YANG ctype error in #807; if it works for you, hit merge and it will join this PR.

File: src/apps/lwaftr/binding_table.lua [hunk 9]
```diff
@@ -158,9 +155,9 @@ function BindingTable:remove_softwire_entry(entry_key_blob)
    self.softwires:remove(entry.key)
 end
 
-local lookup_key = softwire_key_t()
 function BindingTable:lookup(ipv4, port)
-   local psid = self:lookup_psid(ipv4, port)
+   local lookup_key = ffi.new(self.softwires.entry_type).key
```
Comment (wingo) [inline src/apps/lwaftr/binding_table.lua]: This is going to be bad for lookup perf (though we usually use the other interface).  Can we add a FIXME here?

**Review by wingo — APPROVED** [review]

File: src/apps/lwaftr/binding_table.lua (hunk 9 above)
Comment (tsyesika) [reply src/apps/lwaftr/binding_table.lua]: Hm, how about we just create the lookup key when we create the binding table, seems that's what we were doing in the old code. I'm thinking something like this:

```lua
local lookup_key
function BindingTable.new(psid_map, softwires)
   local ret = {
      psid_map = assert(psid_map),
      softwires = assert(softwires),
   }
   lookup_key = ret.softwires.entry_type().key
   return setmetatable(ret, {__index=BindingTable})
end
```
then this can just be reverted to the way it was and I think the code will be effectively the same, right? If I'm missing something more complicated then we can put a FIXME but I'd perfer to fix it now if it's easy.

Comment (wingo) [reply src/apps/lwaftr/binding_table.lua]: Good idea, this looks good to me!

Comment (tsyesika) [reply src/apps/lwaftr/binding_table.lua]: Hokay, i'll fix this and push it in a second :)

**Comment by wingo** [comment]
LGTM.  Let's merge this then then see about merging lwaftr-kristian to lwaftr.
